# PROYECTO FINAL

## SETUP

Instalación e importación librerias

In [1]:
#!pip install pandas==2.2.2
#!pip install ipython-autotime

In [2]:
%load_ext autotime

time: 0 ns (started: 2024-05-06 10:05:19 -03:00)


In [3]:
import pandas as pd
import os


time: 657 ms (started: 2024-05-06 10:05:19 -03:00)


In [4]:
separador = "=" * 60

time: 0 ns (started: 2024-05-06 10:05:19 -03:00)


## IMPORTACIÓN DATASETS ORIGINALES

El primer paso es cargar los csv que contienen la información de cada una de las tablas que vamos a utilizar

In [5]:
# Ubicación de los archivos
loc_orders="https://raw.githubusercontent.com/Andresmup/recursos-analisis-datos/main/14_Datasets_Originales/ecommerce_orders_dataset.csv"
loc_orders_items="https://raw.githubusercontent.com/Andresmup/recursos-analisis-datos/main/14_Datasets_Originales/ecommerce_order_items_dataset.csv"
loc_products="https://raw.githubusercontent.com/Andresmup/recursos-analisis-datos/main/14_Datasets_Originales/ecommerce_products_dataset.csv"
loc_customers="https://raw.githubusercontent.com/Andresmup/recursos-analisis-datos/main/14_Datasets_Originales/ecommerce_customers_dataset.csv"
loc_payments="https://raw.githubusercontent.com/Andresmup/recursos-analisis-datos/main/14_Datasets_Originales/ecommerce_order_payments_dataset.csv"

time: 0 ns (started: 2024-05-06 10:05:19 -03:00)


### Orders

In [6]:
#Importar dataset como Dataframe y previsualizamos
df_orders = pd.read_csv(loc_orders)
df_orders

order_id                       customer_id  \
0      e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1      53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2      47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3      949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4      ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
...                                 ...                               ...   
99436  9c5dedf39a927c1b2549525ed64a053c  39bd1228ee8140590ac3aca26f2dfe00   
99437  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
99438  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
99439  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
99440  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

      order_status order_purchase_timestamp    order_approved_at  \
0        delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1        delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2        delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3        delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4        delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   
...            ...                      ...                  ...   
99436    delivered      2017-03-09 09:54:05  2017-03-09 09:54:05   
99437    delivered      2018-02-06 12:58:58  2018-02-06 13:10:37   
99438    delivered      2017-08-27 14:46:43  2017-08-27 15:04:16   
99439    delivered      2018-01-08 21:28:27  2018-01-08 21:36:21   
99440    delivered      2018-03-08 20:57:30  2018-03-09 11:20:28   

      order_delivered_carrier_date order_delivered_customer_date  \
0              2017-10-04 19:55:00           2017-10-10 21:25:13   
1              2018-07-26 14:31:00           2018-08-07 15:27:45   
2              2018-08-08 13:50:00           2018-08-17 18:06:29   
3              2017-11-22 13:39:59           2017-12-02 00:28:42   
4              2018-02-14 19:46:34           2018-02-16 18:17:02   
...                            ...                           ...   
99436          2017-03-10 11:18:03           2017-03-17 15:08:01   
99437          2018-02-07 23:22:42           2018-02-28 17:37:56   
99438          2017-08-28 20:52:26           2017-09-21 11:24:17   
99439          2018-01-12 15:35:03           2018-01-25 23:32:54   
99440          2018-03-09 22:11:59           2018-03-16 13:08:30   

      order_estimated_delivery_date  
0               2017-10-18 00:00:00  
1               2018-08-13 00:00:00  
2               2018-09-04 00:00:00  
3               2017-12-15 00:00:00  
4               2018-02-26 00:00:00  
...                             ...  
99436           2017-03-28 00:00:00  
99437           2018-03-02 00:00:00  
99438           2017-09-27 00:00:00  
99439           2018-02-15 00:00:00  
99440           2018-04-03 00:00:00  

[99441 rows x 8 columns]

time: 6.28 s (started: 2024-05-06 10:05:20 -03:00)


### Order items

In [7]:
#Importar dataset como Dataframe y previsualizamos
df_order_items = pd.read_csv(loc_orders_items)
df_order_items

order_id  order_item_id  \
0       00010242fe8c5a6d1ba2dd792cb16214              1   
1       00018f77f2f0320c557190d7a144bdd3              1   
2       000229ec398224ef6ca0657da4fc703e              1   
3       00024acbcdf0a6daa1e931b038114c75              1   
4       00042b26cf59d7ce69dfabb4e55b4fd9              1   
...                                  ...            ...   
112645  fffc94f6ce00a00581880bf54a75a037              1   
112646  fffcd46ef2263f404302a634eb57f7eb              1   
112647  fffce4705a9662cd70adb13d4a31832d              1   
112648  fffe18544ffabc95dfada21779c9644f              1   
112649  fffe41c64501cc87c801fd61db3f6244              1   

                              product_id                         seller_id  \
0       4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1       e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2       c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3       7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4       ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   
...                                  ...                               ...   
112645  4aa6014eceb682077f9dc4bffebc05b0  b8bc237ba3788b23da09c0f1f3a3288c   
112646  32e07fd915822b0765e448c4dd74c828  f3c38ab652836d21de61fb8314b69182   
112647  72a30483855e2eafc67aee5dc2560482  c3cfdc648177fdbbbb35635a37472c53   
112648  9c422a519119dcad7575db5af1ba540e  2b3e4a2a3ea8e01938cabda2a3e5cc79   
112649  350688d9dc1e75ff97be326363655e01  f7ccf836d21b2fb1de37564105216cc1   

        shipping_limit_date   price  freight_value  
0       2017-09-19 09:45:35   58.90          13.29  
1       2017-05-03 11:05:13  239.90          19.93  
2       2018-01-18 14:48:30  199.00          17.87  
3       2018-08-15 10:10:18   12.99          12.79  
4       2017-02-13 13:57:51  199.90          18.14  
...                     ...     ...            ...  
112645  2018-05-02 04:11:01  299.99          43.41  
112646  2018-07-20 04:31:48  350.00          36.53  
112647  2017-10-30 17:14:25   99.90          16.95  
112648  2017-08-21 00:04:32   55.99           8.72  
112649  2018-06-12 17:10:13   43.00          12.79  

[112650 rows x 7 columns]

time: 6.22 s (started: 2024-05-06 10:05:26 -03:00)


### Products

In [8]:
#Importar dataset como Dataframe y previsualizamos
df_products = pd.read_csv(loc_products)
df_products

product_id              product_category_name  \
0      1e9e8ef04dbcff4541ed26657ea517e5                         perfumaria   
1      3aa071139cb16b67ca9e5dea641aaa2f                              artes   
2      96bd76ec8810374ed1b65e291975717f                      esporte_lazer   
3      cef67bcfe19066a932b7673e239eb23d                              bebes   
4      9dc1a7de274444849c219cff195d0b71              utilidades_domesticas   
...                                 ...                                ...   
32946  a0b7d5a992ccda646f2d34e418fff5a0                   moveis_decoracao   
32947  bf4538d88321d0fd4412a93c974510e6  construcao_ferramentas_iluminacao   
32948  9a7c6041fa9592d9d9ef6cfe62a71f8c                    cama_mesa_banho   
32949  83808703fc0706a22e264b9d75f04a2e             informatica_acessorios   
32950  106392145fca363410d287a815be6de4                    cama_mesa_banho   

       product_name_lenght  product_description_lenght  product_photos_qty  \
0                     40.0                       287.0                 1.0   
1                     44.0                       276.0                 1.0   
2                     46.0                       250.0                 1.0   
3                     27.0                       261.0                 1.0   
4                     37.0                       402.0                 4.0   
...                    ...                         ...                 ...   
32946                 45.0                        67.0                 2.0   
32947                 41.0                       971.0                 1.0   
32948                 50.0                       799.0                 1.0   
32949                 60.0                       156.0                 2.0   
32950                 58.0                       309.0                 1.0   

       product_weight_g  product_length_cm  product_height_cm  \
0                 225.0               16.0               10.0   
1                1000.0               30.0               18.0   
2                 154.0               18.0                9.0   
3                 371.0               26.0                4.0   
4                 625.0               20.0               17.0   
...                 ...                ...                ...   
32946           12300.0               40.0               40.0   
32947            1700.0               16.0               19.0   
32948            1400.0               27.0                7.0   
32949             700.0               31.0               13.0   
32950            2083.0               12.0                2.0   

       product_width_cm  
0                  14.0  
1                  20.0  
2                  15.0  
3                  26.0  
4                  13.0  
...                 ...  
32946              40.0  
32947              16.0  
32948              27.0  
32949              20.0  
32950               7.0  

[32951 rows x 9 columns]

time: 1.64 s (started: 2024-05-06 10:05:32 -03:00)


### Customers

In [9]:
#Importar dataset como Dataframe y previsualizamos
df_customers = pd.read_csv(loc_customers)
df_customers

customer_id                customer_unique_id  \
0      06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1      18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2      4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3      b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4      4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   
...                                 ...                               ...   
99436  17ddf5dd5d51696bb3d7c6291687be6f  1a29b476fee25c95fbafc67c5ac95cf8   
99437  e7b71a9017aa05c9a7fd292d714858e8  d52a67c98be1cf6a5c84435bd38d095d   
99438  5e28dfe12db7fb50a4b2f691faecea5e  e9f50caf99f032f0bf3c55141f019d99   
99439  56b18e2166679b8a959d72dd06da27f9  73c2643a0a458b49f58cea58833b192e   
99440  274fa6071e5e17fe303b9748641082c8  84732c5050c01db9b23e19ba39899398   

       customer_zip_code_prefix          customer_city customer_state  
0                         14409                 franca             SP  
1                          9790  sao bernardo do campo             SP  
2                          1151              sao paulo             SP  
3                          8775        mogi das cruzes             SP  
4                         13056               campinas             SP  
...                         ...                    ...            ...  
99436                      3937              sao paulo             SP  
99437                      6764        taboao da serra             SP  
99438                     60115              fortaleza             CE  
99439                     92120                 canoas             RS  
99440                      6703                  cotia             SP  

[99441 rows x 5 columns]

time: 5.12 s (started: 2024-05-06 10:05:34 -03:00)


### Payments

In [10]:
#Importar dataset como Dataframe y previsualizamos
df_payments = pd.read_csv(loc_payments)
df_payments

order_id  payment_sequential payment_type  \
0       b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1       a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2       25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3       ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4       42fdf880ba16b47b59251dd489d4441a                   1  credit_card   
...                                  ...                 ...          ...   
103881  0406037ad97740d563a178ecc7a2075c                   1       boleto   
103882  7b905861d7c825891d6347454ea7863f                   1  credit_card   
103883  32609bbb3dd69b3c066a6860554a77bf                   1  credit_card   
103884  b8b61059626efa996a60be9bb9320e10                   1  credit_card   
103885  28bbae6599b09d39ca406b747b6632b1                   1       boleto   

        payment_installments  payment_value  
0                          8          99.33  
1                          1          24.39  
2                          1          65.71  
3                          8         107.78  
4                          2         128.45  
...                      ...            ...  
103881                     1         363.31  
103882                     2          96.80  
103883                     1          47.77  
103884                     5         369.54  
103885                     1         191.58  

[103886 rows x 5 columns]

time: 2.42 s (started: 2024-05-06 10:05:39 -03:00)


## DATOS DATAFRAME

Ahora es momento de analizar como se componen cada uno de los DataFrame, saber cual es su tamaño, ver que tipo de datos tienen ademas de conocer la cantidad de nulos y duplicados.

### Funcion nulos

In [11]:
# Con esta funcion podemos crear una tabla que nos muestre la cantidad de valores nulos y su porcentaje del total de datos
def nulos(df):
    # Calcular la cantidad de valores nulos por columna
    nulos_por_columna = df.isna().sum()
    
    # Calcular el porcentaje de valores nulos por columna
    porcentaje_nulos_por_columna = round(df.isna().sum() / len(df) * 100, 2)
    
    # Crear un DataFrame para almacenar los resultados
    tabla_nulos = pd.DataFrame({'Nulos': nulos_por_columna, 'Porcentaje de Nulos': porcentaje_nulos_por_columna})
    
    print(tabla_nulos)


time: 0 ns (started: 2024-05-06 10:05:41 -03:00)


### Funcion datosDataFrame

In [12]:
def datosDataFrame(df):
    separador = "=" * 60
    
    print(f"{separador}")  # Imprime una línea separadora
    print("Tamaño DataFrame\n")  # Muestra el tamaño del DataFrame
    print(df.shape)  # Tamaño del dataframe
    print(f"{separador}")  # Imprime una línea separadora
    
    print(f"{separador}")  # Imprime una línea separadora
    print("Análisis tipo de datos\n")  # Muestra información sobre los tipos de datos en el DataFrame
    print(df.info())  # Tipos de datos
    print(f"{separador}")  # Imprime una línea separadora
    
    print(f"{separador}")  # Imprime una línea separadora
    print("Recuento de nulos y su porcentaje\n")  # Muestra el número de valores nulos en cada columna del DataFrame y su porcentaje
    nulos(df)  # Análisis de nulos
    print(f"{separador}")  # Imprime una línea separadora

    print(f"{separador}")  # Imprime una línea separadora
    print("Recuento de duplicados\n")  # Muestra el número de filas duplicadas en el DataFrame
    print(df.duplicated().sum())  # Análisis duplicados
    print(f"{separador}")  # Imprime una línea separadora


time: 0 ns (started: 2024-05-06 10:05:41 -03:00)


### Orders

In [13]:
#Aplicamos la función
datosDataFrame(df_orders)

Tamaño DataFrame

(99441, 8)
Análisis tipo de datos

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB
None
Recuento de nulos y su porcentaje

                               Nulos  Porcentaje de Nulos
order_id                           0                 0.00
customer_id                        0                 

### Order items

In [14]:
#Aplicamos la función
datosDataFrame(df_order_items)

Tamaño DataFrame

(112650, 7)
Análisis tipo de datos

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB
None
Recuento de nulos y su porcentaje

                     Nulos  Porcentaje de Nulos
order_id                 0                  0.0
order_item_id            0                  0.0
product_id               0                  0.0
seller_id                0                  0.0
shipping_limit_date      0         

### Products

In [15]:
#Aplicamos la función
datosDataFrame(df_products)

Tamaño DataFrame

(32951, 9)
Análisis tipo de datos

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB
None
Recuento de nulos y su porcentaje

                            Nulos  Porcentaje de Nulos
product_id                      0                 0.00
product_ca

### Customers

In [16]:
#Aplicamos la función
datosDataFrame(df_customers)

Tamaño DataFrame

(99441, 5)
Análisis tipo de datos

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB
None
Recuento de nulos y su porcentaje

                          Nulos  Porcentaje de Nulos
customer_id                   0                  0.0
customer_unique_id            0                  0.0
customer_zip_code_prefix      0                  0.0
customer_city                 0                  0.0
customer_state                0                  0.0
Recuento de duplicados

0
time: 94 ms (started: 2024-

### Payments

In [17]:
#Aplicamos la función
datosDataFrame(df_payments)

Tamaño DataFrame

(103886, 5)
Análisis tipo de datos

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  object 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  object 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB
None
Recuento de nulos y su porcentaje

                      Nulos  Porcentaje de Nulos
order_id                  0                  0.0
payment_sequential        0                  0.0
payment_type              0                  0.0
payment_installments      0                  0.0
payment_value             0                  0.0
Recuento de duplicados

0
time: 78 ms (started: 2024-05-06 10:05:42 -03:00)


## TRANSFORMACION DATAFRAMES

Para asegurar que nuestro analisis no tenga problemas a futuro es importante transformar los tipos de datos, ademas de asegurarse que los datos sean homogeneos.

En aquellos dataframe que hay nulos si la proporcion de nulos no es muy grande es factible eliminar dichas filas.

### Funcion transformar_fechas

In [18]:
#Definimos nuestra función transformar_fechas
def transformar_fechas(df, columnas_fechas): #Los parametros son el dataframe y la lista con las columnas a transformar
    
    df[columnas_fechas] = df[columnas_fechas].apply(pd.to_datetime) #Transforma las columnas de columnas_fechas a datetime
    
    return df

time: 0 ns (started: 2024-05-06 10:05:42 -03:00)


### Funcion transformar_string

In [19]:
#Definimos nuestra función transformar_string

def transformar_string(df, columnas_string): #Los parametros son el dataframe y la lista con las columnas a transformar
    for columna in columnas_string: #Transforma las columnas en columnas_string a string
        df[columna] = df[columna].astype("string")
    return df

time: 0 ns (started: 2024-05-06 10:05:42 -03:00)


### Funcion convertir_mayusculas

In [20]:
#Definimos nuestra función convertir_mayusculas

def convertir_mayusculas(df, columna): #Los parametros son el dataframe y la lista con las columna a transformar
    
    df[columna] = df[columna].str.upper() # Convertir todos los valores de la columna a mayusculas
    
    return df

time: 0 ns (started: 2024-05-06 10:05:42 -03:00)


### Funcion convertir_minusculas

In [21]:
#Definimos nuestra funcion convertir_minusculas

def convertir_minusculas(df, columna): #Los parametros son el dataframe y la lista con las columna a transformar
   
    df[columna] = df[columna].str.lower() # Convertir todos los valores de la columna a minúsculas
    
    return df

time: 16 ms (started: 2024-05-06 10:05:42 -03:00)


### Funcion nulos_eliminar

In [22]:
def nulos_eliminar(df):
    
    print(separador)
    # Porcentaje de nulos en cada columna
    null_percentages = round((df.isnull().sum() / len(df)) * 100,2)

    #Veamos cual es el % de nulos
    print(null_percentages)

    # Filtramos las columnas con menos del 10% de valores nulos
    columns_to_keep = null_percentages[null_percentages <= 10].index

    # Veamos las columnas que cumplen dicha condicion
    print(separador)
    print("Columnas con menos del 10% de valores nulos")
    print()
    for columna in columns_to_keep:
        print(columna)
  
    # Eliminamos las filas con nulos solo en las columnas seleccionadas
    
    df.dropna(subset=columns_to_keep, inplace=True)
    
    #Chequeo cantidad de null
    print(separador)
    nulos(df)
        

time: 0 ns (started: 2024-05-06 10:05:42 -03:00)


### Orders

In [23]:
nulos_eliminar(df_orders)

order_id                         0.00
customer_id                      0.00
order_status                     0.00
order_purchase_timestamp         0.00
order_approved_at                0.16
order_delivered_carrier_date     1.79
order_delivered_customer_date    2.98
order_estimated_delivery_date    0.00
dtype: float64
Columnas con menos del 10% de valores nulos

order_id
customer_id
order_status
order_purchase_timestamp
order_approved_at
order_delivered_carrier_date
order_delivered_customer_date
order_estimated_delivery_date
                               Nulos  Porcentaje de Nulos
order_id                           0                  0.0
customer_id                        0                  0.0
order_status                       0                  0.0
order_purchase_timestamp           0                  0.0
order_approved_at                  0                  0.0
order_delivered_carrier_date       0                  0.0
order_delivered_customer_date      0                  0.0
order_

In [24]:
# Lista de columnas que contienen fechas
columnas_fechas_orders = ['order_purchase_timestamp', 'order_approved_at',
                          'order_delivered_carrier_date', 'order_delivered_customer_date',
                          'order_estimated_delivery_date']

#Aplicamos nuestra funcion conversora de fechas
df_orders = transformar_fechas(df_orders, columnas_fechas_orders)

time: 266 ms (started: 2024-05-06 10:05:42 -03:00)


In [25]:
# Lista de columnas a transfomar en string
columnas_string_orders = ["order_id","customer_id","order_status"]

# Seleccionar columnas que pertenecen a la lista
columnas_filtradas = df_orders.columns[df_orders.columns.isin(columnas_string_orders)]

# Crear un nuevo DataFrame con las columnas filtradas
#df_filtrado = df_orders[columnas_filtradas]
df_filtrado = df_orders.loc[:, columnas_filtradas]


# Mostrar el DataFrame filtrado
#df_filtrado

#Aplicamos nuestra funcion conversora de string
df_orders = transformar_string(df_orders, columnas_string_orders)

time: 16 ms (started: 2024-05-06 10:05:42 -03:00)


In [26]:
#Revisamos que los tipos de datos sean correctos
df_orders.info()

<class 'pandas.core.frame.DataFrame'>
Index: 96461 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       96461 non-null  string        
 1   customer_id                    96461 non-null  string        
 2   order_status                   96461 non-null  string        
 3   order_purchase_timestamp       96461 non-null  datetime64[ns]
 4   order_approved_at              96461 non-null  datetime64[ns]
 5   order_delivered_carrier_date   96461 non-null  datetime64[ns]
 6   order_delivered_customer_date  96461 non-null  datetime64[ns]
 7   order_estimated_delivery_date  96461 non-null  datetime64[ns]
dtypes: datetime64[ns](5), string(3)
memory usage: 6.6 MB
time: 0 ns (started: 2024-05-06 10:05:42 -03:00)


In [27]:
#Aplicamos la conversion a minusculas
df_orders = convertir_minusculas(df_orders, "order_status")

time: 16 ms (started: 2024-05-06 10:05:42 -03:00)


### Order items

In [28]:
# Lista de columnas que contienen fechas
columnas_fechas_order_items = ['shipping_limit_date']

#Aplicamos nuestra funcion conversora de fechas
df_order_items = transformar_fechas(df_order_items, columnas_fechas_order_items)

time: 78 ms (started: 2024-05-06 10:05:42 -03:00)


In [29]:
# Lista de columnas a transfomar en string
columnas_string_order_items = ["order_id","order_item_id","product_id","seller_id"]

#Aplicamos nuestra funcion conversora de string
df_order_items = transformar_string(df_order_items, columnas_string_order_items)

time: 63 ms (started: 2024-05-06 10:05:42 -03:00)


### Products

In [30]:
nulos_eliminar(df_products)

product_id                    0.00
product_category_name         1.85
product_name_lenght           1.85
product_description_lenght    1.85
product_photos_qty            1.85
product_weight_g              0.01
product_length_cm             0.01
product_height_cm             0.01
product_width_cm              0.01
dtype: float64
Columnas con menos del 10% de valores nulos

product_id
product_category_name
product_name_lenght
product_description_lenght
product_photos_qty
product_weight_g
product_length_cm
product_height_cm
product_width_cm
                            Nulos  Porcentaje de Nulos
product_id                      0                  0.0
product_category_name           0                  0.0
product_name_lenght             0                  0.0
product_description_lenght      0                  0.0
product_photos_qty              0                  0.0
product_weight_g                0                  0.0
product_length_cm               0                  0.0
product_height_c

In [31]:
# Lista de columnas a transfomar en string
columnas_string_products = ["product_id","product_category_name"]

#Aplicamos nuestra funcion conversora de string
df_products = transformar_string(df_products, columnas_string_products)

time: 16 ms (started: 2024-05-06 10:05:43 -03:00)


In [32]:
#Revisamos que los tipos de datos sean correctos
df_products.info()

<class 'pandas.core.frame.DataFrame'>
Index: 32340 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32340 non-null  string 
 1   product_category_name       32340 non-null  string 
 2   product_name_lenght         32340 non-null  float64
 3   product_description_lenght  32340 non-null  float64
 4   product_photos_qty          32340 non-null  float64
 5   product_weight_g            32340 non-null  float64
 6   product_length_cm           32340 non-null  float64
 7   product_height_cm           32340 non-null  float64
 8   product_width_cm            32340 non-null  float64
dtypes: float64(7), string(2)
memory usage: 2.5 MB
time: 0 ns (started: 2024-05-06 10:05:43 -03:00)


In [33]:
#Aplicamos la conversion a minusculas
df_products = convertir_minusculas(df_products, "product_category_name")

time: 0 ns (started: 2024-05-06 10:05:43 -03:00)


### Customers

In [34]:
# Lista de columnas a transfomar en string
columnas_string_customers = ["customer_id","customer_unique_id","customer_zip_code_prefix","customer_city","customer_state"]

#Aplicamos nuestra funcion conversora de string
df_customers = transformar_string(df_customers, columnas_string_customers)

time: 47 ms (started: 2024-05-06 10:05:43 -03:00)


In [35]:
#Revisamos que los tipos de datos sean correctos
df_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  string
 1   customer_unique_id        99441 non-null  string
 2   customer_zip_code_prefix  99441 non-null  string
 3   customer_city             99441 non-null  string
 4   customer_state            99441 non-null  string
dtypes: string(5)
memory usage: 3.8 MB
time: 15 ms (started: 2024-05-06 10:05:43 -03:00)


In [36]:
#Aplicamos la conversion a minusculas
df_customers = convertir_minusculas(df_customers, "customer_city")

#Aplicamos la conversion a mayusculas
df_customers = convertir_mayusculas(df_customers, "customer_state")

time: 47 ms (started: 2024-05-06 10:05:43 -03:00)


### Payments

In [37]:
# Lista de columnas a transfomar en string
columnas_string_order_payments = ["order_id","payment_type"]

#Aplicamos nuestra funcion conversora de string
df_order_payments = transformar_string(df_payments, columnas_string_order_payments)

#Aplicamos la conversion a minusculas
df_order_payments = convertir_minusculas(df_order_payments, "payment_type")

time: 31 ms (started: 2024-05-06 10:05:43 -03:00)


## NUEVOS DATAFRAMES

A partir de los dataset originales y despues de trabajar con las columnas que los componen vamos a generar dos nuevos DataFrames que exportaremos a csv para poder utilizarlos y generar un reporte con visualizaciones

In [38]:
# Crear la carpeta 'dataset_generados' si no existe
if not os.path.exists('dataset_generados'):
    os.makedirs('dataset_generados')

time: 0 ns (started: 2024-05-06 10:05:43 -03:00)


In [39]:
#Realizamos los merge
df_merge1 = pd.merge(df_orders, df_payments, how = "inner", on= "order_id")
df_merge2 = pd.merge(df_merge1, df_order_items, how = "inner", on= "order_id")
df_merge3 = pd.merge(df_merge2, df_customers, how = "inner", on= "customer_id")

time: 578 ms (started: 2024-05-06 10:05:43 -03:00)


In [40]:
columnas_a_borrar=['order_approved_at','order_delivered_carrier_date',
                   'order_delivered_customer_date','order_estimated_delivery_date',
                   'payment_sequential','payment_installments','shipping_limit_date',
                  'customer_zip_code_prefix','customer_unique_id','seller_id','order_item_id'
                  ]
df_merge3.drop(columnas_a_borrar, axis=1, inplace=True) #Indicamos el nombre de las columnas y que vamos a mantener los cambios


time: 47 ms (started: 2024-05-06 10:05:43 -03:00)


In [41]:
# Extracción de atributos para la fecha de compra - Año y Mes
df_merge3['order_purchase_year'] = df_merge3['order_purchase_timestamp'].apply(lambda x: x.year)
df_merge3['order_purchase_month'] = df_merge3['order_purchase_timestamp'].apply(lambda x: x.month)
df_merge3['order_purchase_month_name'] = df_merge3['order_purchase_timestamp'].apply(lambda x: x.strftime('%b'))

# Extraer atributos para la fecha de compra - Día de la semana
df_merge3['order_purchase_day'] = df_merge3['order_purchase_timestamp'].apply(lambda x: x.day)
df_merge3['order_purchase_dayofweek_name'] = df_merge3['order_purchase_timestamp'].apply(lambda x: x.strftime('%a'))


# Extracción de atributos para la fecha de compra: hora y momento del día
df_merge3['order_purchase_hour'] = df_merge3['order_purchase_timestamp'].apply(lambda x: x.hour)
hours_bins = [-0.1, 6, 12, 18, 23]
hours_labels = ['Madrugada', 'Mañana', 'Tarde', 'Noche']
df_merge3['order_purchase_time_day'] = pd.cut(df_merge3['order_purchase_hour'], hours_bins, labels=hours_labels)

time: 2 s (started: 2024-05-06 10:05:43 -03:00)


In [42]:
#Realizamos un merge
df_merge4 = pd.merge(df_merge3, df_products, how = "inner", on= "product_id")
df_merge4

order_id                       customer_id  \
0       e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1       e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
2       e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
3       128e10d95713541c87cd1a2e48201934  a20e8105f23924cd00833fd87daa0831   
4       0e7e841ddf8f8f2de2bad69267ecfbcf  26c7ac168e1433912a51b924fbd34d34   
...                                  ...                               ...   
113385  e8fd20068b9f7e6ec07068bb7537f781  609b9fb8cad4fe0c7b376f77c8ab76ad   
113386  e8fd20068b9f7e6ec07068bb7537f781  609b9fb8cad4fe0c7b376f77c8ab76ad   
113387  cfa78b997e329a5295b4ee6972c02979  a2f7428f0cafbc8e59f20e1444b67315   
113388  9c5dedf39a927c1b2549525ed64a053c  39bd1228ee8140590ac3aca26f2dfe00   
113389  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

       order_status order_purchase_timestamp payment_type  payment_value  \
0         delivered      2017-10-02 10:56:33  credit_card          18.12   
1         delivered      2017-10-02 10:56:33      voucher           2.00   
2         delivered      2017-10-02 10:56:33      voucher          18.59   
3         delivered      2017-08-15 18:29:31  credit_card          37.77   
4         delivered      2017-08-02 18:24:47  credit_card          37.77   
...             ...                      ...          ...            ...   
113385    delivered      2017-08-10 21:21:07  credit_card         748.24   
113386    delivered      2017-08-10 21:21:07  credit_card         748.24   
113387    delivered      2017-12-20 09:52:41  credit_card          71.04   
113388    delivered      2017-03-09 09:54:05  credit_card          85.08   
113389    delivered      2018-03-08 20:57:30   debit_card          86.86   

                              product_id   price  freight_value  \
0       87285b34884572647811a353c7ac498a   29.99           8.72   
1       87285b34884572647811a353c7ac498a   29.99           8.72   
2       87285b34884572647811a353c7ac498a   29.99           8.72   
3       87285b34884572647811a353c7ac498a   29.99           7.78   
4       87285b34884572647811a353c7ac498a   29.99           7.78   
...                                  ...     ...            ...   
113385  0df37da38a30a713453b03053d60d3f7  356.00          18.12   
113386  0df37da38a30a713453b03053d60d3f7  356.00          18.12   
113387  3d2c44374ee42b3003a470f3e937a2ea   55.90          15.14   
113388  ac35486adb7b02598c182c2ff2e05254   72.00          13.08   
113389  006619bbed68b000c8ba3f8725d5409e   68.50          18.36   

              customer_city  ... order_purchase_hour  order_purchase_time_day  \
0                 sao paulo  ...                  10                   Mañana   
1                 sao paulo  ...                  10                   Mañana   
2                 sao paulo  ...                  10                   Mañana   
3                 sao paulo  ...                  18                    Tarde   
4                 sao paulo  ...                  18                    Tarde   
...                     ...  ...                 ...                      ...   
113385            sao paulo  ...                  21                    Noche   
113386            sao paulo  ...                  21                    Noche   
113387     campo do tenente  ...                   9                   Mañana   
113388  sao jose dos campos  ...                   9                   Mañana   
113389                 lapa  ...                  20                    Noche   

        product_category_name product_name_lenght  product_description_lenght  \
0       utilidades_domesticas                40.0                       268.0   
1       utilidades_domesticas                40.0                       268.0   
2       utilidades_domesticas                40.0                       268.0   
3       utilidades_domesticas                40.0                       268.0   

time: 531 ms (started: 2024-05-06 10:05:45 -03:00)


## EXPORTAR DATASETS LIMPIOS

## ANALISIS DATAFRAMES

Despues de tranformar los DataFrame podemos realizar un análisis de los datos que contienen, para poder extraer información relevante (insights).

### Funcion analisis_data_frames_round

In [43]:
#Definimos una función para obtener insights relevantes de nuestros dataframes
def analisis_data_frames_round(df): #Como parametro tenemos los dataframes
  print(separador)
  print("Análisis variables cuantitativas\n")
  columnas_numericas = df.select_dtypes(include=['int', 'float']).columns #Extraemos las columnas numericas
  if not columnas_numericas.empty: #Aplicamos un if en caso de que no tenga columnas numéricas el dataframe
      descriptivas=df[columnas_numericas].describe()
      descriptivas_redondeadas = descriptivas.round(2)
      print(descriptivas_redondeadas) #Usamos el describe para aquellas columnas cuantitativas
  print(separador)
  print(separador)
  print("Análisis variables cualitativas\n")
  columnas_string = df.select_dtypes(include=['string']).columns #Extraemos las columnas string
  if not columnas_string.empty: #Aplicamos un if en caso de que no tenga columnas del tipo string el dataframe
    print(df[columnas_string].describe()) #Usamos el describe para aquellas columnas cualitativas

time: 0 ns (started: 2024-05-06 10:05:46 -03:00)


### Funcion tipos_porcentaje

In [44]:
# Con esta funcion podemos crear una tabla que nos muestre la cantidad de valores y su porcentaje
def tipos_porcentaje(df):
    # Calcular la cantidad de valores por columna
    valores =df.value_counts()
    
    total_valores=len(df)
    
    # Calcular el porcentaje de valores por columna
    porcentaje_valores = round(valores / total_valores * 100, 2)
    
    # Crear un DataFrame para almacenar los resultados
    tabla_valores = pd.DataFrame({'Cantidad': valores, 'Porcentaje': porcentaje_valores})
    
    print(tabla_valores)


time: 0 ns (started: 2024-05-06 10:05:46 -03:00)


In [45]:
#Aplicamos la función
analisis_data_frames_round(df_products)

Análisis variables cuantitativas

       product_name_lenght  product_description_lenght  product_photos_qty  \
count             32340.00                    32340.00            32340.00   
mean                 48.48                      771.49                2.19   
std                  10.25                      635.12                1.74   
min                   5.00                        4.00                1.00   
25%                  42.00                      339.00                1.00   
50%                  51.00                      595.00                1.00   
75%                  57.00                      972.00                3.00   
max                  76.00                     3992.00               20.00   

       product_weight_g  product_length_cm  product_height_cm  \
count          32340.00           32340.00           32340.00   
mean            2276.96              30.85              16.96   
std             4279.29              16.96              13.64   
min

In [46]:
#Analizamos las categorias
tipos_porcentaje(df_products.product_category_name)

                               Cantidad  Porcentaje
product_category_name                              
cama_mesa_banho                    3029        9.37
esporte_lazer                      2867        8.87
moveis_decoracao                   2657        8.22
beleza_saude                       2444        7.56
utilidades_domesticas              2335        7.22
...                                 ...         ...
fashion_roupa_infanto_juvenil         5        0.02
casa_conforto_2                       5        0.02
pc_gamer                              3        0.01
seguros_e_servicos                    2        0.01
cds_dvds_musicais                     1         0.0

[73 rows x 2 columns]
time: 15 ms (started: 2024-05-06 10:05:46 -03:00)


Podemos observar que claramente los productos mas vendidos son aquellos relacionados con el hogar (articulos del habitación, baño, utilies del hogar).
Ademas vemos que en la mayoria de los casos con tener solo una foto en la descripción de los productos es suficiente para atraer compradores.

In [47]:
#Aplicamos la función
analisis_data_frames_round(df_customers)

Análisis variables cuantitativas

Análisis variables cualitativas

                             customer_id                customer_unique_id  \
count                              99441                             99441   
unique                             99441                             96096   
top     06b8999e2fba1a1fbc88172c00ba8bc7  8d50f5eadf50201ccdcedfb9e2ac8455   
freq                                   1                                17   

       customer_zip_code_prefix customer_city customer_state  
count                     99441         99441          99441  
unique                    14994          4119             27  
top                       22790     sao paulo             SP  
freq                        142         15540          41746  
time: 188 ms (started: 2024-05-06 10:05:46 -03:00)


In [48]:
#Realizamos un analisis por regiones de los compradores
tipos_porcentaje(df_customers.customer_state)

                Cantidad  Porcentaje
customer_state                      
SP                 41746       41.98
RJ                 12852       12.92
MG                 11635        11.7
RS                  5466         5.5
PR                  5045        5.07
SC                  3637        3.66
BA                  3380         3.4
DF                  2140        2.15
ES                  2033        2.04
GO                  2020        2.03
PE                  1652        1.66
CE                  1336        1.34
PA                   975        0.98
MT                   907        0.91
MA                   747        0.75
MS                   715        0.72
PB                   536        0.54
PI                   495         0.5
RN                   485        0.49
AL                   413        0.42
SE                   350        0.35
TO                   280        0.28
RO                   253        0.25
AM                   148        0.15
AC                    81        0.08
A

Podemos ver que los clientes se encuentran en 3 Estados principalemente:
- Sao Pablo
- Rio de Janeiro
- Minas Gerais

In [49]:
#Realizamos un analisis por regiones de los compradores
tipos_porcentaje(df_merge4.customer_state)

                Cantidad  Porcentaje
customer_state                      
SP                 47830       42.18
RJ                 14651       12.92
MG                 13234       11.67
RS                  6284        5.54
PR                  5790        5.11
SC                  4161        3.67
BA                  3859         3.4
DF                  2389        2.11
GO                  2319        2.05
ES                  2288        2.02
PE                  1804        1.59
CE                  1482        1.31
PA                  1071        0.94
MT                  1067        0.94
MS                   826        0.73
MA                   808        0.71
PB                   614        0.54
RN                   555        0.49
PI                   546        0.48
AL                   437        0.39
SE                   384        0.34
TO                   330        0.29
RO                   275        0.24
AM                   167        0.15
AC                    92        0.08
A

Podemos ver que los clientes se encuentran en 3 Estados principalmente:
- Sao Pablo
- Rio de Janeiro
- Minas Gerais

In [50]:
tipos_porcentaje(df_merge4.order_status)

              Cantidad  Porcentaje
order_status                      
delivered       113383       99.99
canceled             7        0.01
time: 16 ms (started: 2024-05-06 10:05:46 -03:00)


Podemos observar que claramente el servicio de ecommerce tiene una clientela, un servicio de atención y entrega de productos muy fiable ya que del total de productos ordenados el 99.99% de ellos llegan a destino antes de ser cancelados

In [51]:
#Analizamos las categorias
tipos_porcentaje(df_merge4.product_category_name)

                                               Cantidad  Porcentaje
product_category_name                                              
cama_mesa_banho                                   11649       10.27
beleza_saude                                       9761        8.61
esporte_lazer                                      8731         7.7
moveis_decoracao                                   8553        7.54
informatica_acessorios                             7897        6.96
...                                                 ...         ...
portateis_cozinha_e_preparadores_de_alimentos        14        0.01
cds_dvds_musicais                                    14        0.01
pc_gamer                                              9        0.01
fashion_roupa_infanto_juvenil                         7        0.01
seguros_e_servicos                                    2         0.0

[73 rows x 2 columns]
time: 16 ms (started: 2024-05-06 10:05:46 -03:00)


Podemos observar que claramente los productos mas vendidos son aquellos relacionados con el hogar (articulos del habitación, baño, utilies del hogar).
Ademas vemos que en la mayoria de los casos con tener solo una foto en la descripción de los productos es suficiente para atraer compradores.

In [52]:
#Podemos ver como son los tipos de pagos
tipos_porcentaje(df_merge4.payment_type)

              Cantidad  Porcentaje
payment_type                      
credit_card      83706       73.82
boleto           22047       19.44
voucher           6012         5.3
debit_card        1625        1.43
time: 16 ms (started: 2024-05-06 10:05:46 -03:00)


Claramente el método de pago mas elegído es la tarjeta de crédito.